File: embedded/src/alarm/manager.go
package alarm

import (
	"context"
	"time"

	"alarm_project/embedded/src/hardware"
	"alarm_project/embedded/src/log"
)

// AlarmManager handles the execution of an alarm
type AlarmManager struct {
	led    *hardware.LED
	buzzer *hardware.Buzzer
	alarm  Alarm
}

// NewAlarmManager creates a new alarm manager instance
func NewAlarmManager(alarm Alarm) *AlarmManager {
	return &AlarmManager{
		led:    hardware.NewLED(),
		buzzer: hardware.NewBuzzer(),
		alarm:  alarm,
	}
}

// Run executes the alarm sequence
func (r *AlarmManager) Run(ctx context.Context) {
	log.Info("Alarm triggered: %s", r.alarm.Name)
	
	startTime := time.Now()
	ticker := time.NewTicker(50 * time.Millisecond)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			elapsed := time.Since(startTime)
			progress := float64(elapsed) / float64(time.Minute*time.Duration(r.alarm.Length))
			
			if progress >= 1.0 {
				log.Debug("Alarm '%s' intensity curve complete, starting wake-up tune", r.alarm.Name)
				go r.buzzer.PlayWakeUpTune(ctx)
				r.led.SetIntensity(r.alarm.IntensityCurve.EndIntensity)
				
				<-ctx.Done()
				r.cleanup()
				log.Info("Alarm '%s' completed", r.alarm.Name)
				return
			}
			
			intensity := applyIntensityCurve(r.alarm.IntensityCurve, progress)
			r.led.SetIntensity(intensity)
			
		case <-ctx.Done():
			log.Info("Alarm '%s' interrupted", r.alarm.Name)
			r.cleanup()
			return
		}
	}
}

// cleanup ensures all hardware is properly reset
func (r *AlarmManager) cleanup() {
	r.led.SetIntensity(0)
}

File: embedded/src/alarm/manager.py
import time
from alarm.intensity import get_intensity
from log.log import log

class AlarmManager:
    def __init__(self, led, buzzer, button):
        self.led = led
        self.buzzer = buzzer
        self.button = button

    def run_alarm(self, alarm):
        log(f"Alarm '{alarm.name}' triggered!")
        duration = alarm.length * 60  # minutes to seconds
        start = alarm.intensity_curve.start_intensity
        end = alarm.intensity_curve.end_intensity
        curve_type = alarm.intensity_curve.curve
        hyper = alarm.intensity_curve.hyper_parameter
        points = alarm.intensity_curve.control_points
        t0 = time.time()
        cancelled = False
        while True:
            elapsed = time.time() - t0
            t = min(1.0, elapsed / duration)
            intensity = int(get_intensity(curve_type, start, end, t, hyper, points))
            self.led.set_intensity(intensity)
            if self.button.is_pressed():
                log("Alarm cancelled by button press.")
                cancelled = True
                break
            if elapsed >= duration:
                break
            time.sleep(0.1)
        self.led.set_intensity(end)
        if not cancelled:
            self.buzzer.play_melody()
        self.led.off()
        self.buzzer.stop()
        log(f"Alarm '{alarm.name}' finished.")
